### PR TITLE
CS-2634/demo_app

### PR DESCRIPTION
**Pull request by vasil-s**
Create a demo application that showcases using Digital Factory programatically. Included functionality:
* Authenticating with DF.
* Creating a new project in DF.
* Uploading a file to a project.
* Adding a comment to a project.
* Searching for projects.
* Submitting print jobs.
* Fetching running print jobs.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "digital-factory-sdk",
+  "version": "1.0.0",
+  "description": "Try out the Ultimaker Digital Factory API",
+  "main": "index.js",
+  "scripts": {
+    "start": "tsc && node dist/main.js"
+  },
+  "author": "Simon Edwards",
```
Comment (sajanv88) [inline package.json:9]: supposed to be Ultimaker?

Comment (vasil-s) [reply package.json:9]: Copy past, will fix.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "digital-factory-sdk",
```
Comment (sajanv88) [inline package.json:2]: Also, I would suggest renaming this to `digital-factory-API-examples`

File: src/main.ts [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        print("");
+        print(`    ${signInUrl}`);
+        print("");
+        return new Promise<void>((resolve, reject) => {
+            this._signInCompleteResolve = resolve;
+        });
+	}
+
+	private async _handleRequest(req: IncomingMessage, res: ServerResponse): Promise<void> {
+		if (!req.url) {
```
Comment (sajanv88) [inline src/main.ts:70]:
```suggestion
		if (!req.url || !this._pkceVerifier) return;
```

File: src/main.ts [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+            return;
+        }
+		if (!this._pkceVerifier) {
+            return;
+        }
+
+		const url = new URL(`http://localhost:${CALLBACK_SERVER_PORT}${req.url}`);
+
+		const code = url.searchParams.get('code');
```
Comment (sajanv88) [inline src/main.ts:79]:
```suggestion
		const code = url.searchParams.get('code');
	        const state = url.searchParams.get('state');
	        if( !code || ( !state || state != this._state ) ) return;

```

**Review by sajanv88 — COMMENTED** [review]
umm, I feel it may be hard to understand for others.

Since it is a javascript, We could write simple functions that would make it a lot easier to understand.

can we make it simple, please?

For example:
```
// How to get an access token. 
async function getAccessToken () { ... };
const accessToken = await getAccessToken()

// How to create project in digital factory using the access token. 
async createProject (options) { 
const response = await fetch(`${API_ROOT_URL}/cura/v1/projects`, {...options, headers: {  'Authorization': `Bearer ${accessToken}`
} }); 
return await response.json();
 }



```

File: typescript/package.json [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  "name": "digital-factory-sdk",
+  "version": "1.0.0",
+  "description": "Try out the Ultimaker Digital Factory API",
+  "main": "main.ts",
+  "scripts": {
+    "start": "tsc && node dist/main.js"
+  },
+  "author": "Ultimaker <software-cloud@ultimaker.com>",
+  "license": "ICS",
+  "dependencies": {
```
Comment (xli12) [inline typescript/package.json:11]: Are we going to set up dependabot for this repo?

File: typescript/src/main.ts [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+				client_id: CLIENT_ID,
+				redirect_uri: this._redirectUri,
+				grant_type: 'authorization_code',
+				scope: SCOPES,
+				code: code,
+				code_verifier: pkceVerifier,
+			})
+		})
+
+		return tokenResponse.json()
```
Comment (xli12) [inline typescript/src/main.ts:118]: Shall we make eslint as a dev dependency to improve the syntax consistency (e.g. always `;` at the end, the use of single and double quotes, one or two lines between code blocks?)

File: typescript/src/main.ts [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    } else {
+        print("No running print jobs found. Sometimes it takes up to 10 second for new print jobs to show up.\n");
+    }
+
+
+    print("Searching projects.")
+    const projects = await demo.searchProjects();
+    if (projects.length > 0) {
+        print(`Total projects retrieved: ${projects.length}`);
+        print(`First project retrieved: ${JSON.stringify(projects[0])}\n`);
```
Comment (xli12) [inline typescript/src/main.ts:290]: `JSON.stringify(projects[0], null, 4)` to make the printing contents more readable?

**Review by xli12 — CHANGES_REQUESTED** [review]
Shall we set up the linter first?

File: typescript/package.json (hunk 5 above)
Comment (sedwards2009) [reply typescript/package.json:11]: I don't think that is needed here.

File: typescript/src/main.ts (hunk 6 above)
Comment (sedwards2009) [reply typescript/src/main.ts]: I just want this to be as simple as possible, given it is example code.

I've gone through and make it all more consistent, especially the `;`.

File: typescript/src/main.ts (hunk 7 above)
Comment (sedwards2009) [reply typescript/src/main.ts:295]: That does sound like a good idea.

Comment (sedwards2009) [reply typescript/src/main.ts]: done

**Comment by sedwards2009** [comment]
@sajanv88  This is already set up as simple functions/methods which share some fields and are grouped in an object. Loose functions will need either global vars or a ton of parameters.

**Review by xli12 — APPROVED** [review]

**Pull request merged**